core/src/main/java/org/infinispan/distribution/TriangleOrderManager.java

```diff
@@ -0,0 +1,187 @@
+package org.infinispan.distribution;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.infinispan.statetransfer.OutdatedTopologyException;
+import org.infinispan.topology.CacheTopology;
+
+import net.jcip.annotations.GuardedBy;
+
+/**
+ * It manages the order of updates from the primary owner to backup owner.
+ * <p>
+ * It depends on the cache topology id. The primary owner assigns the sequence number to the backup command and then
+ * sends it to the backup owner. In the backup owner, the command awaits until it is its turn to be executed.
+ * <p>
+ * If the command topology id does not match, it throws an {@link OutdatedTopologyException}.
+ * <p>
+ * The sequence order starts with 1 and it is per segment based. This allows segments to be updated concurrently.
+ *
+ * @author Pedro Ruivo
+ * @since 9.0
+ */
+public class TriangleOrderManager {
+
+   private final Map<Integer, SenderSequencer> senderSegmentMap;
+   private final Map<Integer, ReceiverSequencer> receiverSegmentMap;
+   private volatile CacheTopology currentCacheTopology;
+
+   public TriangleOrderManager() {
+      senderSegmentMap = new ConcurrentHashMap<>();
+      receiverSegmentMap = new ConcurrentHashMap<>();
+   }
+
+   public int calculateSegmentId(Object key) {
```

Comment:
I'd rather not expose this in the `TriangleOrderManager`, as both callers already have access to `ClusterDependentLogic`.